central/tlsconfig/tlsconfig_test.go

```diff
@@ -40,3 +48,225 @@ func (s *tlsConfigTestSuite) isCommonNameInCerts(DERs [][]byte, commonName strin
+	tmpDir := s.T().TempDir()
+
+	_, keys, err := createCertChainWithLengthOf(1)
+	s.Require().NoError(err)
+	s.Require().NoError(writeKey(tmpDir, keys[0]))
+
+	badCertFile, err := os.Create(path.Join(tmpDir, TLSCertFileName))
+	s.Require().NoError(err)
+	_, err = badCertFile.WriteString("invalid cert")
+	s.Require().NoError(err)
+	s.Require().NoError(badCertFile.Close())
+
+	_, err = MaybeGetDefaultTLSCertificateFromDirectory(tmpDir)
+	s.ErrorContains(err, "failed to find any PEM data in certificate input")
+}
+
+func (s *tlsConfigTestSuite) TestMaybeGetDefaultTLSCertificate_ShouldReturnAnErrorWhenKeyIsMalformed() {
+	tmpDir := s.T().TempDir()
+
+	certs, _, err := createCertChainWithLengthOf(1)
+	s.Require().NoError(err)
+	s.Require().NoError(writeCerts(tmpDir, certs...))
+
+	badKeyFile, err := os.Create(path.Join(tmpDir, TLSKeyFileName))
+	s.Require().NoError(err)
+	_, err = badKeyFile.WriteString("invalid key")
+	s.Require().NoError(err)
+	s.Require().NoError(badKeyFile.Close())
+
+	_, err = MaybeGetDefaultTLSCertificateFromDirectory(tmpDir)
+	s.ErrorContains(err, "failed to find any PEM data in key input")
+}
+
+func (s *tlsConfigTestSuite) TestMaybeGetDefaultTLSCertificate_ShouldReturnErrorAndWarnTheUserWhenCertsAreInWrongOrder() {
+	tmpDir := s.T().TempDir()
+
+	certs, keys, err := createCertChainWithLengthOf(3)
+	s.Require().NoError(err)
+	s.Require().NoError(writeCerts(tmpDir, certs[1], certs[0], certs[2]))
+	s.Require().NoError(writeKey(tmpDir, keys[0]))
+
+	_, err = MaybeGetDefaultTLSCertificateFromDirectory(tmpDir)
+	s.ErrorContains(err, "private key does not match public key")
+	s.ErrorContains(err, "ensure that the certificate chain is in the correct order")
+}
+
+func createCertChainWithLengthOf(length int) ([]*x509.Certificate, []*ecdsa.PrivateKey, error) {
+	var certs []*x509.Certificate
+	var privateKeys []*ecdsa.PrivateKey
+	for i := 0; i < length; i++ {
+		var parentCert = &x509.Certificate{}
+		var parentKey = &ecdsa.PrivateKey{}
+		if i > 0 {
+			parentCert = certs[i-1]
+			parentKey = privateKeys[i-1]
+		}
+		serialNumber, err := generateSerialNumber()
+		if err != nil {
+			return nil, nil, err
+		}
+		template := &x509.Certificate{
+			SerialNumber: serialNumber,
+			IsCA:         i != length-1,
+			Subject: pkix.Name{
+				CommonName: fmt.Sprintf("Cert %d", i),
+			},
+		}
+		privateKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+		if err != nil {
+			return nil, nil, err
+		}
+		if i == 0 {
+			parentKey = privateKey
+		}
+		certBytes, err := x509.CreateCertificate(rand.Reader, template, parentCert, &privateKey.PublicKey, parentKey)
+		if err != nil {
+			return nil, nil, err
+		}
+		cert, err := x509.ParseCertificate(certBytes)
+		if err != nil {
+			return nil, nil, err
+		}
+		certs = append(certs, cert)
+		privateKeys = append(privateKeys, privateKey)
+	}
+	// inverse the order of the certs so that the leaf is first
+	for i, j := 0, len(certs)-1; i < j; i, j = i+1, j-1 {
```

Comment:
`<rant>`

:woozy_face: this makes my head ache.
Yeah, after 5 minutes I do understand what this does and that it's probably the most efficient way, and that it DTRT in all corner cases, but I really hope the standard library will grow a generic reverse function at some point...
`</rant>`

Reply:
I agree! Just found out that we already have a method (`sliceutils.ReverseInPlace`). Substituted for that one instead of the ad-hoc reverse.